File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.py
import sys

n = int(sys.stdin.readline().rstrip())
element = []
for _ in range(n):
    data = list(map(int, sys.stdin.readline().split()))
    element.append(data)

answer = int(1e9)


def recur(idx, sin, sen, flag):
    global answer
    if idx == n:
        if flag == 0:
            return
        answer = min(answer, abs(sin - sen))
        return
    recur(idx + 1, sin * element[idx][0], sen + element[idx][1], flag + 1)
    recur(idx + 1, sin, sen, flag)


recur(0, 1, 0, 0)
print(answer)

File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	
	static int answer = Integer.MAX_VALUE;
	static int[][] menu;
	static int t;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		st = new StringTokenizer(br.readLine());
		t = Integer.parseInt(st.nextToken());
		menu = new int[t][22];
		
		for (int i = 0; i<t; i++) {
			st = new StringTokenizer(br.readLine());
			menu[i][0] = Integer.parseInt(st.nextToken());
			menu[i][1] = Integer.parseInt(st.nextToken());
		}
		
		recur(1,0,0,0);
		System.out.println(answer);
	}
	public static void recur(int start, int end, int idx, int flag) {
		
		if (idx == t) {
			if (flag == 0) {
				return;
			}
			answer = Math.min(answer, Math.abs(end-start));
			return;
		}
		
		recur(start, end, idx+1,flag);
		recur(start*menu[idx][0], end+menu[idx][1], idx+1,flag+1);
	}
}